File: TodoApp/ViewModels/ItemViewModel.cs
namespace TodoApp.ViewModels {
    using DevExpress.Mvvm;
    using DevExpress.Mvvm.POCO;
    using TodoApp.Data;
    using TodoApp.Model;

    public class ItemViewModel : ISupportParameter {
        public ItemViewModel() {
            // Start listen the ReloadRequired message 
            Messenger.Default.Register<ReloadRequired>(this, OnReloadRequired);
        }
        void OnReloadRequired(ReloadRequired message) {
            if(message.IsNew) ReloadItem(message.Id);
        }
        object ISupportParameter.Parameter {
            get { return Item; }
            set { ReloadItem((int)value); }
        }
        protected IRepository Repository {
            get { return this.GetRequiredService<IRepository>(); }
        }
        void ReloadItem(int id) {
            Item = Repository.LoadItem(id);
        }
        // Current item
        public virtual TodoItem Item {
            get;
            protected set;
        }
        // Modification flag
        public virtual bool HasChanges {
            get;
            protected set;
        }
        // View's Title
        public string Title {
            get {
                if(Item == null)
                    return string.Empty;
                var suffix = (HasChanges ? "*" : string.Empty);
                if(IsNew)
                    return "New task" + suffix;
                else
                    return "Task #" + Item.Id.ToString() + suffix;
            }
        }
        bool IsNew {
            get { return Item.Id == AppModel.NewItemID; }
        }
        protected void OnItemChanged() {
            HasChanges = false;
            this.RaisePropertyChanged(x => x.Title);
            this.RaiseCanExecuteChanged(x => x.Save());
            this.RaiseCanExecuteChanged(x => x.Delete());
        }
        protected void OnHasChangesChanged() {
            this.RaisePropertyChanged(x => x.Title);
            this.RaiseCanExecuteChanged(x => x.Save());
        }
        public void Update() {
            HasChanges = Repository.HasChanges(Item.Id, Item);
        }
        public bool CanSave() {
            return (Item != null) && HasChanges;
        }
        public void Save() {
            bool isNew = IsNew;
            int savedId = Repository.Save(Item);
            if(isNew) {
                // Sending the ReloadRequired message for new item
                Messenger.Default.Send(ReloadRequired.FromNew(savedId));
            }
            else {
                // Sending the ReloadRequired message for the specific item
                Messenger.Default.Send(ReloadRequired.FromId(savedId));
            }

            HasChanges = false;
        }
        public bool CanDelete() {
            return (Item != null) && !IsNew;
        }
        public void Delete() {
            if(Repository.Delete(Item.Id)) {
                // Sending the ReloadRequired message for all items
                Messenger.Default.Send(ReloadRequired.All);
                Close();
            }
        }
        protected INavigationService NavigationService {
            get { return this.GetService<INavigationService>(); }
        }
        public void Close() {
            var document = NavigationService.Current as IDocument;
            NavigationService.GoBack();
            if(document != null)
                document.Close(true);
        }
    }
}

File: VB/ViewModels/ItemViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports TodoApp.Data
Imports TodoApp.Model

Namespace TodoApp.ViewModels
    Public Class ItemViewModel
        Implements ISupportParameter
        Public Sub New()
            ' Start listen the ReloadRequired message 
            Messenger.Default.Register(Me, New System.Action(Of ReloadRequired)(AddressOf OnReloadRequired))
        End Sub
        Sub OnReloadRequired(ByVal message As ReloadRequired)
            If message.IsNew Then ReloadItem(message.Id)
        End Sub
        Property Parameter As Object Implements ISupportParameter.Parameter
            Get
                Return Item
            End Get
            Set(ByVal value As Object)
                ReloadItem(CInt(value))
            End Set
        End Property
        Protected ReadOnly Property Repository As IRepository
            Get
                Return GetRequiredService(Of IRepository)()
            End Get
        End Property
        Sub ReloadItem(ByVal id As Integer)
            Item = Repository.LoadItem(id)
        End Sub
        ' Current item
        Public Overridable Property Item As TodoItem
        ' Modification flag
        Public Overridable Property HasChanges As Boolean
        ' View's Title
        Public ReadOnly Property Title As String
            Get
                If Item Is Nothing Then Return String.Empty
                Dim suffix = If(HasChanges, "*", String.Empty)
                If IsNew Then
                    Return "New task" & suffix
                Else
                    Return "Task #" & Item.Id.ToString() & suffix
                End If
            End Get
        End Property
        ReadOnly Property IsNew As Boolean
            Get
                Return Item.Id = AppModel.NewItemID
            End Get
        End Property
        Protected Sub OnItemChanged()
            HasChanges = False
            RaisePropertyChanged(Function(x) x.Title)
            RaiseCanExecuteChanged(Sub(x) x.Save())
            RaiseCanExecuteChanged(Sub(x) x.Delete())
        End Sub
        Protected Sub OnHasChangesChanged()
            RaisePropertyChanged(Function(x) x.Title)
            RaiseCanExecuteChanged(Sub(x) x.Save())
        End Sub
        Public Sub Update()
            HasChanges = Repository.HasChanges(Item.Id, Item)
        End Sub
        Public Function CanSave() As Boolean
            Return Item IsNot Nothing AndAlso HasChanges
        End Function
        Public Sub Save()
            Dim isNew As Boolean = Me.IsNew
            Dim savedId As Integer = Repository.Save(Item)
            If isNew Then
                ' Sending the ReloadRequired message for new item
                Messenger.Default.Send(ReloadRequired.FromNew(savedId))
            Else
                ' Sending the ReloadRequired message for the specific item
                Messenger.Default.Send(ReloadRequired.FromId(savedId))
            End If
            HasChanges = False
        End Sub
        Public Function CanDelete() As Boolean
            Return Item IsNot Nothing AndAlso Not IsNew
        End Function
        Public Sub Delete()
            If Repository.Delete(Item.Id) Then
                ' Sending the ReloadRequired message for all items
                Messenger.Default.Send(ReloadRequired.All)
                Close()
            End If
        End Sub
        Protected ReadOnly Property NavigationService As INavigationService
            Get
                Return GetService(Of INavigationService)()
            End Get
        End Property
        Public Sub Close()
            Dim document = TryCast(NavigationService.Current, IDocument)
            NavigationService.GoBack()
            If document IsNot Nothing Then document.Close(True)
        End Sub
    End Class
End Namespace
